Local Polykey Node Discovery

amydevs opened the pull request:
### Description

This PR integrates MDNS into the NodeConnectionManager to allow for the discovery of nodes on the local network, as well as disable holepunching for connections to local network addresses.

This PR should only focus on the Step 1 of #537 spec, as modifying the node graph to accept multiple IP addresses will require ACL changes that default disable the sharing of local addresses with seed nodes, as well as expansion to support multiple IP addresses.

This PR will encapsulate MDNS in the NodeConnectionManager, as well as make methods such as `getConnection` to be able to query MDNS for known hosts through the newly added `findNodesAll` and `findNodesLocal` methods. `findNodesAll` is a method that calls both the original `findNode` and `findNodesLocal` to aggregate a list of all NodeAddresses. This is notably only used in connection-related methods rather than those related to Kademlia (NodeGraph).

When the NodeConnectionManager is started, MDNS will be started, with a Polykey service registered, and a query for Polykey services started.

When the NodeConnectionManager is stopped, MDNS will be stopped alongside all pending queries, and deregistering all registered services.

Current scopes: `external` and `local`.

### Issues Related
<!-- List all issues fixed by this PR. -->
* Related #537 

### Tasks
<!-- 
  List all tasks to be done by this PR.
  If a task is no longer required, add a strikethrough (including the checkbox):
  - ~~[ ] 3. ...~~ - being completed in #...
-->
- [x] 1. Integrate MDNS into NodeConnectionManager
- [x] 2. Separate MDNS and Kademlia Behaviour
- [x] 3. Update NodeConnectionManager Methods to Support Multiple IP Addresses
  - [x]  `getConnectionWithAddress`
  - [x] `getConnection` 
- [x] 4. Select a well-known port and multicast address for Polykey usage.
- [x] 5. Modify Network Domain to Support Link-Local Addresses with a Interface ID Suffix.
- [x] 6. Creating of connections should happen in stages, where local addresses should be tried for first before external addresses 

### Final checklist
<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

* [x] Domain specific tests
* [x] Full tests
* [x] Updated inline-comment documentation
* [x] Lint fixed
* [x] Squash and rebased
* [x] Sanity check the final build

amydevs committed:
feat: initial @matrixai/mdns integration

fix: changed NodeAddress type and added ability for `NodeConnectionManager.findNodes` to use MDNS

chore: updated @matrixiai/mdns to 1.1.2

fix: temporarily removed IPv6 addresses from MDNS NodeAddress data

fix: `NodeConnectionManager.findNodes` has been separated into 3 methods to not have NodeGraph-related methods interfere with MDNS-based discovery

fix: paramaritized the enabling of the MDNS stack within
NodeConnectionManager and disabled use of MDNS in all unrelated
NodeConnectionManager tests

fix: added back IPv6 addresses in MDNS NodeAddress data

fix: NCM getMultiConnection was failing when more than one NodeAddress
was provided

fix: `NodeConnectionManager.seednodes.test.ts` now includes scopes for node addresses

fix: renamed findNodeLocal and findNodeAll to be less ambiguous

chore: inline documentation for `NodeConnectionManager.findNodeLocal`

feat: created `agentMdnsGroups` and `agneMdnsPort` MDNS configuration in `config.defaultsSystem`

chore: bump `@matrixai/mdns` to `v1.1.6`

feat: better handling of disabling iceProm for `local` `NodeAddress`s in `NodeConnectionManager.establishSingleConnection`

fix: client handler input types for `NodesFind` reverted

fix: errors are now thrown if a connection is attempted with no addresses provided

fix `networkUtils.isHost` is now able to validate link-local addresses

fix: temporarily disabled connecting to link-local addresses with `NodeConnection`

chore: bump `@matrixai/mdns` to `v1.2.0`

fix: reduce `any` typecasts in `NodeConnectionManager.findNodeLocal`

fix: broken jest tests due to mandatory `scopes` property of `NodeAddress`

[ci-skip]

CMCDragonkai commented on tests/nodes/NodeConnectionManager.general.test.ts:
```diff
@@ -206,6 +207,7 @@ describe(`${NodeConnectionManager.name} general test`, () => {
     await nodeConnectionManager.start({
       host: localHost as Host,
+      enableMdns: false,
     });
```
Why have it disabled ever? Why not always have it on and fully encapsulated?

amydevs replied:
i disabled it for tests that didn't need MDNS, because some of those tests would fail expecting for a certain node to not be found.

CMCDragonkai replied:
Can we make this a bit more robust? So that it works regardless, so less manual configuration. How would I know when mdns should be enabled or not, remember less configuration is better.

CMCDragonkai replied:
Ideas? Can we make it just work no matter the situation? If it really is indeed a sort of toggle, it should be `mdns: true/false` I think. But better to just always be enabled, we don't have a situation where it doesn't need it.

Unless dependency injection makes sense here? Thus requiring an external thing to create it first and then inject into it, that would make it a required dependency.

amydevs replied:
i think it would make sense for seednodes to have it disabled, maybe injected dep would make sense

amydevs replied:
this is no longer relevant, due to MDNS being an injected optional dependency

amydevs committed:
fix: cleaned up `tests/nodes/NodeConnectionManager.mdns.test.ts`

[ci-skip]

tegefaulkes commented on src/client/types.ts:
```diff
@@ -103,20 +103,31 @@ type NodeIdMessage = {
+  port: number;
   hostname: string;
   usageCount: number;
   timeout: number;
```
This can be simplified, keep the `AddressMesage` type as it was, just whatever needs a bunch of them can include an `Array<AddressMessage>`.

amydevs replied:
Resolved:
```ts
// Nodes messages

type NodeIdMessage = {
  nodeIdEncoded: NodeIdEncoded;
};

type AddressMessage = {
  host: string;
  port: number;
};

type NodeAddressMessage = NodeIdMessage & AddressMessage;

type NodesFindMessage = {
  addresses: Array<AddressMessage>;
};

type NodesGetMessage = NodeAddressMessage &  { bucketIndex: number };

type NodesAddMessage = NodeAddressMessage & {
  force?: boolean;
  ping?: boolean;
};

type NodeConnectionMessage = NodeAddressMessage & {
  hostname: string;
  usageCount: number;
  timeout: number;
};
```

tegefaulkes commented on src/nodes/NodeConnectionManager.ts:
```diff
@@ -407,12 +416,18 @@ class NodeConnectionManager {
     ipv6Only = false,
+    enableMdns = true,
+    mdnsGroups = ['224.0.0.250', 'ff02::fa17'] as Array<Host>,
+    mdnsPort = 64023 as Port,
```
Defaults here should come from the config if it's defined there. I think all the start parameters need to be fixed.

amydevs replied:
This is no longer relevant due to MDNS being an injected optional dep

tegefaulkes commented on src/nodes/NodeConnectionManager.ts:
```diff
@@ -508,6 +546,9 @@ class NodeConnectionManager {
     await this.rpcServer.stop({ force: true });
+    if (this.mdns[running]) {
+      await this.mdns.stop();
+    }
```
Since stop is idempotent I think you can just call stop without checking if its running. 

Usually we only check if something is running prior to stopping if its happening in the background. But in this case the `mdns` could just not be running. So I'm not sure, maybe @CMCDragonkai can weigh in.

amydevs replied:
This is no longer relevant due to MDNS being an injected optional dep

tegefaulkes commented on src/nodes/NodeConnectionManager.ts:
```diff
@@ -893,8 +935,10 @@ class NodeConnectionManager {
+          if (iceProm != null) {
           iceProm.cancel('Connection was established');
           await iceProm;
+          }
```
This can simplify.

```suggestion
          iceProm?.cancel('Connection was established');
          await iceProm;
```

tegefaulkes commented on src/nodes/NodeConnectionManager.ts:
```diff
@@ -651,21 +692,21 @@ class NodeConnectionManager {
-    // If there was no address provided then we need to find it.
-    if (address == null) {
+    // If there were no addresses provided then we need to find them.
+    if (addresses == null) {
```
Since `addresses` is an array now, you need to check if it's an empty array and handle that. It should throw an error as invalid usage. 

It's probably checked somewhere else so likely won't cause a problem. But as the saying goes, fail early and fail often.

amydevs replied:
```ts
// If there were no addresses provided then we need to find them.
    if (addresses == null || addresses.length === 0) {
      // Find the node
      addresses = await this.findNodeAll(targetNodeId, undefined, ctx);
      if (addresses.length === 0) throw new nodesErrors.ErrorNodeGraphNodeIdNotFound();
    }
```

tegefaulkes commented on src/nodes/NodeConnectionManager.ts:
```diff
@@ -700,19 +741,19 @@ class NodeConnectionManager {
+    addresses: Array<NodeAddress>,
     @context ctx: ContextTimed,
   ): Promise<ConnectionAndTimer> {
     const targetNodeIdString = targetNodeId.toString() as NodeIdString;
```
Here too, empty array shouldn't be allowed so check and throw.

amydevs replied:
```ts
if (addresses.length === 0) {
      throw new nodesErrors.ErrorNodeConnectionManagerNodeAddressRequired();
    }
```

tegefaulkes commented on src/nodes/NodeConnectionManager.ts:
```diff
@@ -1173,6 +1217,99 @@ class NodeConnectionManager {
+            host: host as any,
+            port: service.port as any,
+            scopes
+          });
```
A lot of `as any` here, This is a bit of a pet peve for me so try not to do it. If you have to cast then cast it to the type you need.

amydevs replied:
i need to make the types in mdns itself to not use Opaue types for the service detail i think

amydevs replied:
```ts
if (this.mdns != null) {
      const service = this.mdns.networkServices.get(mdnsUtils.toFqdn({ name: nodesUtils.encodeNodeId(targetNodeId), type: "polykey", protocol: "udp" }));
      if (service != null) {
        for (const host_ of service.hosts) {
          let host: string;
          switch (this.quicSocket.type) {
            case 'ipv4':
              if (quicUtils.isIPv4(host_)) host = host_;
              else if (quicUtils.isIPv4MappedIPv6(host_)) host = quicUtils.fromIPv4MappedIPv6(host_);
              else continue;
              break;
            case 'ipv6':
              if (quicUtils.isIPv6(host_)) host = host_;
              else continue;
              break;
            case 'ipv4&ipv6':
              host = host_;
              break;
            default:
              continue;
          }
          const scopes: Array<'local' | 'external'> = ['local'];
          if (externalAddress?.host === host && externalAddress.port === service.port) {
            scopes.push('external');
            concatExternal = false;
          }
          addresses.push({
            host: host as Host,
            port: service.port as Port,
            scopes
          });
          this.logger.debug(
            `found address for ${nodesUtils.encodeNodeId(targetNodeId)} at ${
              host
            }:${service.port}`,
          );
        }
      }
    }
```

tegefaulkes commented on src/nodes/types.ts:
```diff
@@ -6,9 +6,12 @@ import type { Host, Hostname, Port } from '../network/types';
 
+type NodeAddressScope = 'local' | 'external';
+
 type NodeAddress = {
   host: Host | Hostname;
   port: Port;
+  scopes: Array<NodeAddressScope>;
 };
```
Adding in the scopes field makes this a little clunky. But this seems fine for now. We can go over this again with the node graph expansion.

tegefaulkes commented on tests/nodes/NodeConnectionManager.mdns.test.ts:
```diff
@@ -0,0 +1,193 @@
+      seedNodes: undefined,
+    });
+
+    const serviceProm = promise();
```
We need to fully deconstruct the promise now.

amydevs committed:
feat: moved `MDNS` into `PolykeyAgent` to be passed as a optional injected dep to `NodeConnectionManager`

fix: `PolyKeyAgent` now correctly injects `mdns` into the constructor of `NodeConnectionManager`

fix: await mdns.start

fix: `MDNS`ids are now determined by first 16 bits of node id

[ci-skip]

CMCDragonkai commented on src/config.ts:
```diff
@@ -246,6 +248,25 @@ const config = {
+     * The default value has been selected as the decimal decimal representation of `fa17`,
+     * which resembles the latin word `fait`.
+     */
+    agentMdnsPort: 64023,
```
The prefix for this should just be `mdnsGroups` and `mdnsPort` because you're setting them up in `mdns` group in the `PolykeyAgentOptions`.

CMCDragonkai commented on src/nodes/agent/handlers/NodesHolePunchMessageSend.ts:
```diff
@@ -103,6 +103,7 @@ class NodesHolePunchMessageSend extends UnaryHandler<
+          scopes: ['external'],
         });
       }
     });
```
Is this affected by #555 @tegefaulkes?

tegefaulkes replied:
It might be, we do send host and port, but I'm not sure the scope is relevant to our use.

amydevs committed:
fix: renamed `agentMdnsGroups` and `agentMdnsPort` to `mdnsGroups` and `mdnsPort`

[ci-skip]

amydevs committed:
fix: `NodeConnectionManager.findNodeLocal` is now used to initiate MDNS queries, so that connectivity liveness is preserved

fix `NodeConnectionManager.findNodeAll` has been given a default time out of `NodeConnectionManager.connectionConnectTimeoutTIme`

[ci-skip]

amydevs committed:
feat: `NodeConnectionManager.getConnectionWithAddresses` is now separated into multiple stages so that if connecting with `local` `NodeAddress`s succeeds, `external` `NodeAddress`s will be ignored

fix: `NodeConnectionManager.getConnectionWithAddresses` now attempts connections with `local` scoped addresses before `external` scoped addresses

fix: correct typing on `establishMultiConnection` for `ctx` param

[ci-skip]

amydevs committed:
fix: `mdns.groups` and `mdns.port` is now also apart of the `optionsDefaulted` for `PolykeyAgent.start`

[ci-skip]

tegefaulkes commented on src/PolykeyAgent.ts:
```diff
@@ -316,6 +326,15 @@ class PolykeyAgent {
+        logger: logger.getChild(MDNS.name)
+      });
+      mdns.start({
+        id: keyRing.getNodeId().at(0),
```
What is the `id` for exactly? And you're betting it from a byte in the `NodeId`?

I'm not saying it's a problem, just curious.

amydevs replied:
The `id` paramater used so that MDNS knows what packets were sent from itself, and what were by others on the network.
Anyhow, it has been changed to now instead take the first 2 bytes, so that the `MDNS` id has more chance of being unique.

CMCDragonkai replied:
16 bit prefix of node ID? I'm not sure but I think it's better to use a hashing function. However you won't get more than 256 nodes per local context. Can this be expanded to 32bits?

tegefaulkes commented on src/PolykeyAgent.ts:
```diff
@@ -316,6 +326,15 @@ class PolykeyAgent {
       });
+      mdns = new MDNS({
+        logger: logger.getChild(MDNS.name)
+      });
```
I don't see the `mdns` being passed into anything that uses it. The `NodeConnectionManager` needs it right?

amydevs replied:
i forgot to, have added it in

tegefaulkes commented on src/nodes/NodeConnectionManager.ts:
```diff
@@ -783,16 +829,29 @@ class NodeConnectionManager {
     nodeIds: Array<NodeId>,
     addresses: Array<NodeAddress>,
-    ctx: ContextTimed,
+    @context ctx,
```
Should have the type annotation still. `@context ctx: ContextTimed`

Generally I didn't apply the `@timedCancellable` decoration to protected methods,because they should always be called within something that is already `timedCancellable`.

amydevs replied:
have added in

tegefaulkes commented on src/nodes/NodeConnectionManager.ts:
```diff
@@ -1205,6 +1265,124 @@ class NodeConnectionManager {
+  public async findNodeLocal(
+    targetNodeId: NodeId,
+    @context ctx: ContextTimed,
+  ): Promise<Array<NodeAddress>> {
```
If this is a pretty simple operation then we can specify a default timeout on the `@timedCancellable` decorator. 

`findNode` didn't have one because it's a pretty long running and complex search procedure.

amydevs replied:
should this have a default in options?

tegefaulkes replied:
Derive it from the existing options. I think it can just use the normal connection timeout value?

CMCDragonkai replied:
Use the default timeout for finding nodes - shouldn't be different from global vs local.

Also you are using scopes right? Why don't name your suffixes equivalent to the scope keywords?

You have external and local. But not global and local? What about internal? How many more keywords? Did you spec out what an array means for scopes?

amydevs replied:
alright

CMCDragonkai replied:
Re defaults, just use config. It should have relevant defaults for everything.

tegefaulkes reviewed (changes requested):
Some small things but otherwise seems fine.

Don't forget to lint, I saw an unused import somewhere and linting should catch that.

If all the tests pass then you can merge.

amydevs committed:
chore: lintfix

[ci-skip]

amydevs committed:
fix: set a different port in `NodeConnectionManager.mdns.test.ts` to not conflict with existing PolykeyAgents on host

[ci-skip]
